src/geometry/sphere.cc

```diff
@@ -0,0 +1,29 @@
+#include "sphere.h"
+#include "ray.h"
+#include <iostream>
+
+Sphere::Sphere(Vertex position, float radius, ColorDbl color)
+    : SceneObject(position), color_(color) {
+  assert(radius > 0);
+  radius_ = radius;
+}
+
+bool Sphere::RayIntersection(Ray& ray, float& z) {
+  Vertex o = ray.start();
+  Vertex cp = position_;
+  Direction l = glm::normalize(ray.end() - ray.start());
+  float r = radius_;
+
+  //float a = glm::dot(l, l); // == 1
+  float b = glm::dot(2.f*l, o-cp);
+  float c = glm::dot(o-cp, o-cp) - sqrt(r);
```

Comment:
'o-cp' is executed three times. Probably just a minor gain to store it in it's own variable but I usually like to do it when I have to do the calculation more than twice.